drivers/gsm/gsm.c

```diff
@@ -0,0 +1,850 @@
+                    size_t outlen = 0;
+                    while ((*(++pos) != '"') && (len--)) {
+                        *outbuf++ = *pos;
+                        outlen++;
+                    }
+                    if (len) {
+                        *outbuf = '\0';
+                        res = outlen;
+                    }
+                    else {
+                        res = -ENOSPC;
+                    }
+                }
+                else {
+                    res = -1;
+                }
+            }
+        }
+    }
+
+    return res;
+}
+
+
+ssize_t gsm_cmd(gsm_t *dev, const char *cmd, uint8_t *buf, size_t len, unsigned timeout)
+{
+    gsm_lock(dev);
+
+    ssize_t res = at_send_cmd_get_lines(&dev->at_dev, cmd, (char *)buf, len,
+                                        true, (uint32_t)US_PER_SEC * timeout);
+
+    gsm_unlock(dev);
+
+    return res;
+}
+
+
+void gsm_print_status(gsm_t *dev)
+{
+    char buf[GSM_AT_LINEBUFFER_SIZE];
+    uint16_t loc;
+    uint32_t cell;
+
+    int res = gsm_get_identification(dev, buf, GSM_AT_LINEBUFFER_SIZE);
+
+    if (res >= 2) {
+        printf("device type %s\n", buf);
+    }
+    else {
+        printf("error reading device type!\n");
+    }
+
+    res = gsm_get_imei(dev, buf, GSM_AT_LINEBUFFER_SIZE);
+    if (res >= 0) {
+        printf("IMEI: \"%s\"\n", buf);
+    }
+    else {
+        printf("error getting IMEI\n");
+    }
+
+    res = gsm_get_simcard_identification(dev, buf, GSM_AT_LINEBUFFER_SIZE);
+    if (res >= 0) {
+        printf("ICCID: \"%s\"\n", buf);
+    }
+    else {
+        printf("error getting ICCID\n");
+    }
+
+    res = gsm_get_imsi(dev, buf, GSM_AT_LINEBUFFER_SIZE);
+    if (res >= 0) {
+        printf("IMSI: \"%s\"\n", buf);
+    }
+    else {
+        printf("error getting IMSI\n");
+    }
+
+    res = gsm_get_registration(dev, &loc, &cell);
+    if (res > 0) {
+        uint8_t tech = -1;
+        printf("registration code: %d\n", res);
+        printf("location id: %X\n", loc);
+        printf("cell id: %X\n", cell);
+
+        res = gsm_get_operator(dev, buf, GSM_AT_LINEBUFFER_SIZE, &tech);
+        if (res >= 0) {
+            if ((char)tech != -1) {
+                printf("registered to \"%s\" (%u)\n", buf, tech);
+            }
+            else {
+                printf("registered to \"%s\"\n", buf);
+            }
+        }
+        else {
+            printf("no operator\n");
+        }
+    }
+    else {
+        printf("not registered\n");
+    }
+
+    int rssi;
+    unsigned ber;
+    res = gsm_get_signal(dev, &rssi, &ber);
+    if (res == 0) {
+        printf(MODULE"RSSI= %ddBm ber=%u%%\n", rssi, ber);
+    }
+    else {
+        printf(MODULE"error getting signal strength\n");
+    }
+}
+
+#ifdef MODULE_AT_URC
+void gsm_register_urc_callback(gsm_t *dev, at_urc_t *urc)
+{
+    if(dev) {
+        gsm_lock(dev);
+
+        at_add_urc(&dev->at_dev, urc);
+
+        gsm_unlock(dev);
+    }
+}
+
+#endif
+
+#ifdef GSM_HAS_THREAD
```

Comment:
Maybe this macro logic could be simplified as right now there is no use for the thread unless AT_URC is being used.